File: src/gestorAplicacion/obras/Publicacion.java
package src.gestorAplicacion.obras;

import gestorAplicacion.prestamo.Prestamo;

import java.io.Serializable;
import java.util.ArrayList;

public abstract class Publicacion implements Serializable {
	//ATRIBUTOS DE CLASE
	private static final long serialVersionUID = 1L;
	public static ArrayList<Publicacion> lista = new ArrayList<>();
	static protected int numeroPublicaciones;
	public static enum Estados  {CIRCULACION, PRESTADO};
	
	//ATRIBUTOS INSTANCIA
	protected int codigo;
	protected String nombre;
	protected short año;
	protected short ejemplar;
	protected Estados estado = Estados.CIRCULACION;
	protected Estanteria estanteria;
	protected Prestamo prestamo;
	
	
	//CONSTRUCTORES
	Publicacion(int codigo, String nombre, short año, short ejemplar) {
	this.codigo= codigo;
	this.nombre=nombre;
	this.año= año;
	this.ejemplar=ejemplar;
	lista.add(this);
	}
	
	public  String mostrarInfo () {
		return "Informacion perteneciente a la publicacion " + this.nombre ;
	}
	
	//METODOS ABSTRACTOS
	  // muestra informacion detallada de la publicacion, es decir sus atributos
	public abstract String mostrarUbicacion();  //muestra la estanteria y con ello la ubicacion donde se encuentra la publicacion
	
	
	//METODOS NORMALES
	public static void eliminarPublicacion(Publicacion p) { // elimina un registro de Publicacion
		lista.remove(lista.indexOf(p));
	}
	public static String mostrarRegistros() {   // muestra todos los registros creados de publicacion
		String c="Publicaciones creadas: "+"\n";
		for (int i = 0; i < lista.size(); i++) {
			if(lista.get(i) instanceof Libro ) {c = c + i+"."+"Libro "+ lista.get(i).nombre + "  Codigo (CP) "+ lista.get(i).codigo+ "\n";	}
			else if(lista.get(i) instanceof Revista) {c = c + i+"."+"Revista "+ lista.get(i).nombre + "  Codigo (CP) "+ lista.get(i).codigo+ "\n";}
			else if(lista.get(i) instanceof Folleto) {c = c + i+"."+"Folleto "+ lista.get(i).nombre + "  Codigo (CP) "+ lista.get(i).codigo+ "\n";}
		}
		return c;
	}
	
	public boolean verificarPrestado() {
		boolean b= true;                                            //Verifica que la publicacion este en circulacion es decir no prestada, arroja true si esta prestado
		if(this.getEstado()== Estados.CIRCULACION) {
			 b= false;
		}
		return b;
	}
	

	
	
	//GETTERS Y SETTERS
	public long getCodigo() {
		return codigo;
	}
	public void setCodigo(int codigo) {
		this.codigo = codigo;
	}
	public String getNombre() {
		return nombre;
	}
	public void setNombre(String nombre) {
		this.nombre = nombre;
	}
	public short getAño() {
		return año;
	}
	public void setAño(short año) {
		this.año = año;
	}
	public short getEjemplar() {
		return ejemplar;
	}
	public void setEjemplar(short ejemplar) {
		this.ejemplar = ejemplar;
	}
	public Estados getEstado() {
		return estado;
	}
	public void setEstado(Estados estado) {
		this.estado = estado;
	}
	public static int getNumeroPublicaciones() {
		return numeroPublicaciones;
	}
	public static void setNumeroPublicaciones(int numeroPublicaciones) {
		Publicacion.numeroPublicaciones = numeroPublicaciones;
	}
	public static ArrayList<Publicacion> getLista() {
		return lista;
	}
	public static void setLista(ArrayList<Publicacion> lista) {
		Publicacion.lista = lista;
	}
	public Estanteria getEstanteria() {
		return estanteria;
	}
	public void setEstanteria(Estanteria estanteria) {
		this.estanteria = estanteria;
	}
}

File: practica2-python/src/gestorAplicacion/obras/Publicacion.py
from enum import Enum
from abc import ABC, abstractmethod

from gestorAplicacion.obras.Estanteria import Estanteria

#CLASE ABSTRACTA
class Estado(Enum):
    circulacion = "CIRCULACION"
    prestado = "PRESTADO"

class Publicacion(ABC) : 

    #Atributos de clase
    _lista = []
    numeroPublicaciones = 0

    #Constructor
    def __init__(self, codigo=0,nombre="...",año=0,ejemplar=0,estado=Estado.circulacion) :
        self.codigo = codigo
        self.nombre = nombre
        self.año= año
        self.ejemplar = ejemplar
        self.estado = estado
        self.estanteria=None
        self.prestamo=None
        Publicacion.numeroPublicaciones+=1
        Publicacion._lista.append(self)

    #Metodos
    def asignarEstanteria(self,numEstanteria):
        for x in Estanteria.getLista():
            if x.getNumero() == numEstanteria:
                self.setEstanteria(x)
                x.getPublicaciones().append(self)
                break

    def mostrarInfo(self):
        return f'INFORMACION DE LA PUBLICACION: \n Codigo: {self.getCodigo()} \n Nombre: {self.getNombre()} \n Ejemplar: {self.getEjemplar()}'

    @classmethod
    def eliminarPublicacion(cls,p):
        from gestorAplicacion.obras.Libro import Libro
        from gestorAplicacion.obras.Revista import Revista
        from gestorAplicacion.obras.Folleto import Folleto
        if isinstance(p,Libro):
            Libro.getLibro().remove(p)
        elif isinstance(p,Folleto):
            Folleto.getFolleto().remove(p)
        elif isinstance(p,Revista):
            Revista.getRevista().remove(p)
        Publicacion._lista.remove(p)
        del p

    @classmethod
    def buscarPublicacion(cls, Codigop):
        
        for x in Publicacion.getLista():
            if x.getCodigo() == Codigop:
                publicacion = x
                break
        return publicacion

    @classmethod
    def mostrarRegistros(cls):
        from gestorAplicacion.obras.Libro import Libro
        from gestorAplicacion.obras.Revista import Revista
        from gestorAplicacion.obras.Folleto import Folleto
        c = "PUBLICACIONES CREADAS   "+ "\n"
        for i,x in enumerate(Publicacion._lista) :
            if isinstance(x,Libro):
                c=c + str(i) + f'. Libro {x.getNombre()} Codigo(CP) {x.getCodigo()} \n'
            elif isinstance(x,Revista):
                c=c + str(i) + f'. Revista {x.getNombre()} Codigo(CP) {x.getCodigo()} \n'
            elif isinstance(x,Folleto):
                c=c + str(i) + f'. Folleto {x.getNombre()} Codigo(CP) {x.getCodigo()} \n'
        return c
    
    def verificarPrestado(self):
        if self.getEstado() == Estado.prestado:
            b=True
        else :
            b=False
        return b

    @abstractmethod
    def mostrarUbicacion(self):
        pass
    
    
    @classmethod
    def getLista(cls):
        return Publicacion._lista
    @classmethod
    def setLista(cls,lista):
        Publicacion._lista = lista

    @classmethod
    def getNumeroPublicaciones(cls):
        return Publicacion.numeroPublicaciones

    def getCodigo(self):
        return self.codigo
    def setCodigo(self,a):
        self.codigo = a

    def getNombre(self):
        return self.nombre
    def setNombre(self,a):
        self.nombre = a

    def getAño(self):
        return self.año 
    def setAño(self,a):
        self.año= a

    def getEjemplar(self):
        return self.ejemplar
    def setEjemplar(self,a):
        self.ejemplar=a

    def getEstado(self):
        return self.estado
    def setEstado(self,a):
        self.estado = a

    def getEstanteria(self):
        return self.estanteria
    def setEstanteria(self, a):
        self.estanteria = a

    def getPrestamo(self):
        return self.prestamo
    def setPrestamo(self,p):
        self.prestamo = p

